jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,828 @@
+    runtime_type =
+      "java::" + type_from_array->substr(1, type_from_array->length() - 2);
+  }
+  if(!symbol_table.has_symbol(runtime_type))
+    return {}; // Fall back to compile-time type if runtime type was not found
+  const auto &replacement_class_type =
+    to_java_class_type(symbol_table.lookup(runtime_type)->type);
+  return replacement_class_type;
+}
+
+/// Given a JSON representation of an array and a type inferred from the type of
+/// a containing array, get the element type by removing the leading '['.
+/// Types for arrays are stored in the format "[Lmy.package.name.ClassName;".
+/// In this case, the returned value would be "Lmy.package.name.ClassName;".
+/// \p type_from_array would only have a value if this array is stored within
+/// another array, i.e.\ within a ClassName[][].
+/// Keeping track of array types in this way is necessary to assign generic
+/// arrays with no compile-time types.
+/// \param json: JSON representation of an array. If it contains a `@type`
+///   field, this takes priority over \p type_from_array.
+/// \param type_from_array: may contain a type name from a containing array.
+/// \return if the type of an array was given, the type of its elements.
+static optionalt<std::string> element_type_from_array_type(
+  const jsont &json,
+  const optionalt<std::string> &type_from_array)
+{
+  if(has_type(json))
+  {
+    const std::string &json_array_type = json["@type"].value;
+    PRECONDITION(json_array_type.find('[') == 0);
+    return json_array_type.substr(1);
+  }
+  else if(type_from_array)
+  {
+    PRECONDITION(type_from_array->find('[') == 0);
+    return type_from_array->substr(1);
+  }
+  return {};
+}
+
+/// Entry point of the recursive deterministic assignment algorithm.
+/// \param expr: expression to assign a deterministic value to. In the case of
+///   the entry point, this is either a pointer to a struct, or an expression
+///   corresponding to a Java primitive.
+/// \param json: a JSON representation of the deterministic value to assign.
+/// \param type_from_array: if `expr` was found as an element of an array,
+///   the element type of this array.
+/// \param info: references used throughout the recursive algorithm.
+void assign_from_json_rec(
+  const exprt &expr,
+  const jsont &json,
+  const optionalt<std::string> &type_from_array,
+  det_creation_infot &info);
+
+/// One of the base cases (primitive case) of the recursion.
+/// For characters, the encoding in \p json is assumed to be UTF-8.
+/// See \ref assign_from_json_rec.
+static void assign_primitive_from_json(
+  const exprt &expr,
+  const jsont &json,
+  code_blockt &block)
+{
+  if(json.is_null()) // field is not mentioned in json, leave as default value
+    return;
+  if(expr.type() == java_boolean_type())
+  {
+    if(json.is_true())
+      block.add(code_assignt{expr, true_exprt{}});
+    else
+      block.add(code_assignt{expr, false_exprt{}});
```

Comment:
I tried this before, sadly it doesn't work because C++ wants the expressions in a ternary to have _exactly_ the same type. The types here are different (`true_exprt` and `false_exprt`). The way to get around this to write the whole thing in one statement would be to use an immediately invoked lambda and specify its return type, but in this case this really wouldn't make it more readable:
```
    block.add(code_assignt{expr, [&]() -> exprt {
                             if(json.is_true())
                               return true_exprt{};
                             return false_exprt{};
                           }()});
 ```

Reply:
Maybe slightly nicer:
```
    const exprt boolean_expr = [&]() -> exprt {
      if(json.is_true())
        return true_exprt{};
      return false_exprt{};
    }();
    block.add(code_assignt{expr, boolean_expr});
 ```
Any preference?

Reply:
I see. Don't change it then.